File: mobile/androidApp/src/main/kotlin/dev/centraid/android/kit/Rooms.kt
package dev.centraid.android.kit

import androidx.activity.compose.BackHandler
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.rememberUpdatedState
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.heading
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import centraid.screen.v1.Autosave
import centraid.screen.v1.SearchField
import centraid.screen.v1.StatusLine as StatusLineMessage
import dev.centraid.android.theme.centraidColor
import dev.centraid.android.theme.centraidType

/**
 * THE ROOMS (DESIGN.md "The seven rooms"; K5, #1029).
 *
 * A screen is one of these and supplies content and copy; the room owns the
 * header, the back or close control, search, the status host and the band
 * slot. A screen that hand-rolls any of those is a finding, not a variant.
 * `HomeRoom`, `SystemPlace` and `StageRoom` are Home's, Settings' and the
 * lightbox's and are not in this kit yet.
 */

/**
 * The header's one trailing action (≤1 by construction): an icon key from the
 * catalog, and the words a screen reader says for it.
 */
public data class RoomAction(
    val iconKey: String,
    val label: String,
    val testTag: String = "room-action",
    val onPress: () -> Unit,
)

/** Search on an [AppPlace]: the kit `SearchField` and its two events. */
public data class RoomSearch(
    val field: SearchField,
    val placeholder: String,
    val onTerm: (String) -> Unit,
    val onClose: () -> Unit,
    val label: String = KitWords.SEARCH,
    val closeLabel: String = KitWords.CLOSE_SEARCH,
)

/**
 * `AppPlace`: an app's root. `AppHeader` (mark + name + ≤1 trailing action),
 * the search field under it while [search] is open, [content], the status line
 * and the app's band at the foot — a render prop, so the band stays the app's.
 */
@Composable
public fun AppPlace(
    app: String,
    title: String,
    modifier: Modifier = Modifier,
    meta: String = "",
    trailing: RoomAction? = null,
    search: RoomSearch? = null,
    status: String = "",
    band: (@Composable () -> Unit)? = null,
    content: @Composable () -> Unit,
) {
    Column(modifier.fillMaxSize().testTag("$app-place")) {
        Row(
            Modifier.fillMaxWidth().padding(start = KitGeometry.GUTTER, end = 8.dp, top = 8.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            AppMark(appId = app, size = 28.dp, modifier = Modifier.clearAndSetSemantics { })
            Column(Modifier.weight(1f).padding(start = 10.dp)) {
                RoomTitle(title)
                if (meta.isNotEmpty()) {
                    Text(meta, style = centraidType("annotLabel"), color = centraidColor("textSoft"))
                }
            }
            if (trailing != null) IconKey(trailing.iconKey, trailing.label, trailing.testTag, onPress = trailing.onPress)
        }
        if (search != null && search.field.open_) {
            CentraidSearchField(
                field = search.field,
                placeholder = search.placeholder,
                onTerm = search.onTerm,
                onClose = search.onClose,
                label = search.label,
                closeLabel = search.closeLabel,
            )
        }
        Box(Modifier.weight(1f).fillMaxWidth()) { content() }
        StatusLine(status)
        band?.invoke()
    }
}

/**
 * `PushedPage`: content a member pushed into. `PlaceHeader` with back to the
 * NAMED parent ([parentTitle] is what the back key says), the title, ≤1
 * trailing action; the status line at the foot. No band.
 */
@Composable
public fun PushedPage(
    title: String,
    parentTitle: String,
    onBack: () -> Unit,
    modifier: Modifier = Modifier,
    trailing: RoomAction? = null,
    status: String = "",
    content: @Composable () -> Unit,
) {
    Column(modifier.fillMaxSize().testTag("pushed-page")) {
        Row(
            Modifier.fillMaxWidth().padding(start = 4.dp, end = 8.dp, top = 4.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Row(
                Modifier
                    .heightIn(min = KitGeometry.ROW_MIN)
                    .clip(RoundedCornerShape(KitGeometry.RADIUS))
                    .clickable(onClick = onBack)
                    .padding(horizontal = 8.dp)
                    .testTag("room-back")
                    .clearAndSetSemantics {
                        contentDescription = "${KitWords.BACK} to $parentTitle"
                        role = Role.Button
                    },
                verticalAlignment = Alignment.CenterVertically,
            ) {
                CentraidIcon(iconKey = "ChevronLeft", tint = centraidColor("textSoft"), size = 18.dp)
                Text(
                    parentTitle,
                    style = centraidType("annotLabelOn"),
                    color = centraidColor("textSoft"),
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                )
            }
            Spacer(Modifier.weight(1f))
            if (trailing != null) IconKey(trailing.iconKey, trailing.label, trailing.testTag, onPress = trailing.onPress)
        }
        RoomTitle(title, Modifier.padding(horizontal = KitGeometry.GUTTER, vertical = 4.dp))
        Box(Modifier.weight(1f).fillMaxWidth()) { content() }
        StatusLine(status)
    }
}

/**
 * `EditorRoom`: full screen, autosave, CLOSE = DONE (#1015 D3), band hidden.
 *
 * The close key and the system back gesture both call [onClose] (the shell's
 * pop). [onDeparted] runs when the room leaves composition by ANY road — close,
 * back, Home — and is where the route calls its bridge's `departed()`, so
 * unsaved words are saved on the way out. The autosave status line is hosted
 * under the header, never as a button: there is no Save.
 */
@Composable
public fun EditorRoom(
    title: String,
    status: Autosave?,
    onClose: () -> Unit,
    onDeparted: () -> Unit,
    modifier: Modifier = Modifier,
    closeLabel: String = KitWords.DONE,
    trailing: RoomAction? = null,
    content: @Composable ColumnScope.() -> Unit,
) {
    val departed = rememberUpdatedState(onDeparted)
    DisposableEffect(Unit) { onDispose { departed.value() } }
    BackHandler(onBack = onClose)
    Column(modifier.fillMaxSize().testTag("editor-room")) {
        Row(
            Modifier.fillMaxWidth().padding(start = 4.dp, end = 8.dp, top = 4.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            IconKey("X", closeLabel, "room-close", bordered = false, onPress = onClose)
            Column(Modifier.weight(1f).padding(horizontal = 4.dp)) {
                if (title.isNotEmpty()) {
                    Text(
                        title,
                        style = centraidType("smallStrong"),
                        color = centraidColor("text"),
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis,
                        modifier = Modifier.semantics { heading() },
                    )
                }
                AutosaveStatus(status)
            }
            if (trailing != null) IconKey(trailing.iconKey, trailing.label, trailing.testTag, onPress = trailing.onPress)
        }
        Column(Modifier.weight(1f).fillMaxWidth(), content = content)
    }
}

/** The one ink button a [SheetRoom] may carry. */
public data class SheetPrimary(
    val label: String,
    val destructive: Boolean = false,
    val testTag: String = "sheet-primary",
    val onPress: () -> Unit,
)

/**
 * `SheetRoom`: the grabber (Material's drag handle), a title carrying the
 * noun, [content], the status line, and ≤1 ink button. A destructive primary
 * is the OUTLINED `net` button, never a filled one.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
public fun SheetRoom(
    title: String,
    onDismiss: () -> Unit,
    modifier: Modifier = Modifier,
    primary: SheetPrimary? = null,
    status: String = "",
    content: @Composable ColumnScope.() -> Unit,
) {
    ModalBottomSheet(onDismissRequest = onDismiss, containerColor = centraidColor("bg")) {
        Column(modifier.fillMaxWidth().navigationBarsPadding().padding(bottom = 16.dp).testTag("sheet-room")) {
            if (title.isNotEmpty()) RoomTitle(title, Modifier.padding(horizontal = KitGeometry.GUTTER, vertical = 8.dp))
            content()
            StatusLine(status)
            if (primary != null) {
                InkButton(
                    label = primary.label,
                    destructive = primary.destructive,
                    testTag = primary.testTag,
                    modifier = Modifier.padding(horizontal = KitGeometry.GUTTER, vertical = 8.dp),
                    onPress = primary.onPress,
                )
            }
        }
    }
}

/**
 * A full-width button. Ink (`accent` fill) for the one act a sheet offers;
 * a destructive act is outlined in `net` with `net` words and no fill
 * (the hand-off's "destructive outline").
 */
@Composable
public fun InkButton(
    label: String,
    modifier: Modifier = Modifier,
    destructive: Boolean = false,
    testTag: String = "kit-ink-button",
    onPress: () -> Unit,
) {
    val shape = RoundedCornerShape(KitGeometry.RADIUS)
    Box(
        modifier
            .fillMaxWidth()
            .heightIn(min = KitGeometry.ROW_MIN)
            .clip(shape)
            .then(
                if (destructive) {
                    Modifier.border(KitGeometry.HAIRLINE, centraidColor("net"), shape)
                } else {
                    Modifier.background(centraidColor("accent"))
                },
            )
            .clickable(onClick = onPress)
            .testTag(testTag)
            .semantics { role = Role.Button },
        contentAlignment = Alignment.Center,
    ) {
        Text(
            label,
            style = centraidType("control"),
            color = centraidColor(if (destructive) "net" else "onAccent"),
        )
    }
}

/** A room's title: the `title` rung, a heading to a screen reader. */
@Composable
internal fun RoomTitle(title: String, modifier: Modifier = Modifier) {
    Text(
        title,
        style = centraidType("title"),
        color = centraidColor("text"),
        maxLines = 2,
        overflow = TextOverflow.Ellipsis,
        modifier = modifier.semantics { heading() },
    )
}

/** THE STATUS HOST: one quiet line, drawn only when there is something to say. */
@Composable
public fun StatusLine(status: String, modifier: Modifier = Modifier) {
    if (status.isEmpty()) return
    Text(
        status,
        style = centraidType("annotLabel"),
        color = centraidColor("textSoft"),
        modifier = modifier
            .fillMaxWidth()
            .padding(horizontal = KitGeometry.GUTTER, vertical = 6.dp)
            .testTag("room-status"),
    )
}

/**
 * THE KIT `StatusLine` MESSAGE: one clause and at most one verb ("Undo"). A
 * refusal is `net` ink and the core's sentence; [onAct] is the screen's own
 * `StatusActed` event. Drawn only when there is a sentence.
 */
@Composable
public fun StatusLine(line: StatusLineMessage?, modifier: Modifier = Modifier, onAct: () -> Unit) {
    if (line == null || line.sentence.isEmpty()) return
    Row(
        modifier
            .fillMaxWidth()
            .padding(horizontal = KitGeometry.GUTTER, vertical = 4.dp)
            .testTag("room-status-line"),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Text(
            line.sentence,
            style = centraidType("annotLabel"),
            color = centraidColor(if (line.refused) "net" else "textSoft"),
            modifier = Modifier.weight(1f),
        )
        if (line.action_label.isNotEmpty()) {
            QuietButton(line.action_label, testTag = "room-status-act", onPress = onAct)
        }
    }
}

/**
 * AN ICON-ONLY KEY: a 44dp target, a 36dp plate, one catalog mark, and the
 * label as its whole description (the mark itself says nothing).
 */
@Composable
public fun IconKey(
    iconKey: String,
    label: String,
    testTag: String,
    modifier: Modifier = Modifier,
    bordered: Boolean = true,
    onPress: () -> Unit,
) {
    Box(
        modifier
            .size(KitGeometry.ROW_MIN)
            .clip(RoundedCornerShape(KitGeometry.RADIUS))
            .clickable(onClick = onPress)
            .testTag(testTag)
            .clearAndSetSemantics {
                contentDescription = label
                role = Role.Button
            },
        contentAlignment = Alignment.Center,
    ) {
        Box(
            Modifier
                .size(36.dp)
                .then(
                    if (bordered) {
                        Modifier.border(KitGeometry.HAIRLINE, centraidColor("line"), RoundedCornerShape(KitGeometry.RADIUS))
                    } else {
                        Modifier
                    },
                ),
            contentAlignment = Alignment.Center,
        ) {
            CentraidIcon(iconKey = iconKey, tint = centraidColor("textSoft"), size = 18.dp)
        }
    }
}

File: mobile/iosApp/Sources/Kit/Rooms.swift
import SwiftUI

// THE ROOMS (K5; DESIGN.md "The seven rooms (mobile)").
//
// A mobile screen is one of seven rooms and nothing else. The room owns the
// header, the back control, search, the empty/loading/error states (through
// `ReadStateView`), the status host and the gutter; the app supplies content
// and copy. A screen that hand-rolls any of those is a finding, not a variant
// — so a port's view is a room with content in it, and nothing around it.
//
// `HomeRoom` is `HomeView`, `StageRoom` is the lightbox's own, and
// `SystemPlace` is the vault/settings sheets; the four every app port needs
// live here.

/// THE SEARCH SLOT an `AppPlace` opens under its header. The field's state is
/// the machine's `SearchField`; the words are the app's copy.
struct SearchSlot {
    let field: Centraid_Screen_V1_SearchField
    let placeholder: String
    var closeLabel: String = "Close search"
    let onTerm: (String) -> Void
    let onClose: () -> Void
}

/// AN APP'S ROOT: `AppHeader` (mark + name + at most one trailing action), the
/// optional search under it, the content, and the app's band at the foot as a
/// render prop.
///
/// The system navigation bar is hidden: an app place has no back — the band's
/// Home capsule is the way out — and a second header over the app's own would
/// be two grounds for one title. `showsBand` is false for a denied room, which
/// has nowhere in the app to go.
struct AppPlace<Trailing: View, Band: View, Content: View>: View {
    let app: String
    let title: String
    var search: SearchSlot? = nil
    var showsBand: Bool = true
    let trailing: () -> Trailing
    let band: () -> Band
    let content: () -> Content

    @Environment(\.colorScheme) private var scheme

    init(
        app: String,
        title: String,
        search: SearchSlot? = nil,
        showsBand: Bool = true,
        @ViewBuilder trailing: @escaping () -> Trailing,
        @ViewBuilder band: @escaping () -> Band,
        @ViewBuilder content: @escaping () -> Content
    ) {
        self.app = app
        self.title = title
        self.search = search
        self.showsBand = showsBand
        self.trailing = trailing
        self.band = band
        self.content = content
    }

    var body: some View {
        VStack(spacing: 0) {
            HStack(spacing: 10) {
                AppMark(appID: app, size: 28)
                Text(title)
                    .centraidType("title")
                    .foregroundStyle(Theme.color("text", scheme))
                    .lineLimit(1)
                    .accessibilityAddTraits(.isHeader)
                Spacer(minLength: 0)
                trailing()
            }
            .padding(.horizontal, CentraidGeometry.pageMargin)
            .frame(minHeight: 52)
            .accessibilityIdentifier("\(app)-app-header")
            if let search, search.field.open {
                CentraidSearchField(
                    field: search.field,
                    placeholder: search.placeholder,
                    closeLabel: search.closeLabel,
                    identifier: "\(app)-search",
                    onTerm: search.onTerm,
                    onClose: search.onClose
                )
                .padding(.horizontal, CentraidGeometry.pageMargin)
                .padding(.bottom, 8)
            }
            content()
                .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .top)
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .top)
        .background(Theme.color("bg", scheme).ignoresSafeArea())
        .toolbar(.hidden, for: .navigationBar)
        .safeAreaInset(edge: .bottom, spacing: 0) {
            if showsBand {
                band()
                    .background(Theme.color("bg", scheme).ignoresSafeArea(edges: .bottom))
            }
        }
    }
}

/// A PAGE PUSHED FROM A PLACE: `PlaceHeader` with back to the NAMED parent and
/// at most one trailing action. The parent's name is the caller's computed
/// title, never a literal "Back".
struct PushedPage<Trailing: View, Content: View>: View {
    let title: String
    let parentTitle: String
    let onBack: () -> Void
    let trailing: () -> Trailing
    let content: () -> Content

    @Environment(\.colorScheme) private var scheme

    init(
        title: String,
        parentTitle: String,
        onBack: @escaping () -> Void,
        @ViewBuilder trailing: @escaping () -> Trailing,
        @ViewBuilder content: @escaping () -> Content
    ) {
        self.title = title
        self.parentTitle = parentTitle
        self.onBack = onBack
        self.trailing = trailing
        self.content = content
    }

    var body: some View {
        content()
            .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .top)
            .background(Theme.color("bg", scheme).ignoresSafeArea())
            .navigationTitle(title)
            .navigationBarTitleDisplayMode(.inline)
            .navigationBarBackButtonHidden(true)
            .toolbar {
                ToolbarItem(placement: .topBarLeading) {
                    Button(action: onBack) {
                        HStack(spacing: 2) {
                            CentraidIconView(iconKey: "ChevronLeft", tint: Theme.color("text", scheme), size: 18)
                            Text(parentTitle)
                                .centraidType("body")
                                .foregroundStyle(Theme.color("text", scheme))
                                .lineLimit(1)
                                .fixedSize()
                        }
                        .frame(minHeight: CentraidGeometry.targetMinCoarse)
                        .contentShape(Rectangle())
                    }
                    .buttonStyle(.plain)
                    .accessibilityLabel("Back to \(parentTitle)")
                    .accessibilityIdentifier("kit-pushed-back")
                }
                ToolbarItem(placement: .topBarTrailing) { trailing() }
            }
    }
}

/// A FULL-SCREEN EDITOR: autosave, close = done, band hidden, the status line
/// hosted inside (#1015 D3).
///
/// There is no Save. The machine saves on its own debounce; Done only leaves,
/// and leaving is the flush — `onDeparted` runs when the room goes off screen
/// by any route (Done, a pop, Home), and the registry forwards it to the
/// bridge's `departed()`, which sends the machine's `Left` so unsaved words
/// are written without closing the bridge.
struct EditorRoom<Trailing: View, Content: View>: View {
    let title: String
    let status: Centraid_Screen_V1_Autosave
    var closeLabel: String = "Done"
    let onClose: () -> Void
    let onDeparted: () -> Void
    let trailing: () -> Trailing
    let content: () -> Content

    @Environment(\.colorScheme) private var scheme

    init(
        title: String,
        status: Centraid_Screen_V1_Autosave,
        closeLabel: String = "Done",
        onClose: @escaping () -> Void,
        onDeparted: @escaping () -> Void,
        @ViewBuilder trailing: @escaping () -> Trailing,
        @ViewBuilder content: @escaping () -> Content
    ) {
        self.title = title
        self.status = status
        self.closeLabel = closeLabel
        self.onClose = onClose
        self.onDeparted = onDeparted
        self.trailing = trailing
        self.content = content
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            AutosaveStatus(status)
                .padding(.horizontal, CentraidGeometry.pageMargin)
                .padding(.bottom, 4)
            content()
                .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
        }
        .background(Theme.color("bg", scheme).ignoresSafeArea())
        .navigationTitle(title)
        .navigationBarTitleDisplayMode(.inline)
        .navigationBarBackButtonHidden(true)
        .toolbar {
            ToolbarItem(placement: .topBarLeading) {
                Button(action: onClose) {
                    Text(closeLabel)
                        .centraidType("smallStrong")
                        .foregroundStyle(Theme.color("text", scheme))
                        .lineLimit(1)
                        .fixedSize()
                        .padding(.horizontal, 8)
                        .frame(minHeight: CentraidGeometry.targetMinCoarse)
                        .contentShape(Rectangle())
                }
                .buttonStyle(.plain)
                .accessibilityIdentifier("kit-editor-close")
            }
            ToolbarItem(placement: .topBarTrailing) { trailing() }
        }
        .onDisappear(perform: onDeparted)
    }
}

/// THE ONE INK BUTTON a sheet may carry.
struct SheetPrimary {
    let label: String
    var destructive: Bool = false
    let action: () -> Void
}

/// A SHEET: grabber, a title carrying the noun, at most one ink button, and a
/// status line hosted inside. Present it with `.sheet`; the room sets its own
/// grabber and detents.
struct SheetRoom<Content: View>: View {
    let title: String
    var status: String = ""
    var primary: SheetPrimary? = nil
    let content: () -> Content

    @Environment(\.colorScheme) private var scheme

    init(
        title: String,
        status: String = "",
        primary: SheetPrimary? = nil,
        @ViewBuilder content: @escaping () -> Content
    ) {
        self.title = title
        self.status = status
        self.primary = primary
        self.content = content
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            Text(title)
                .centraidType("title")
                .foregroundStyle(Theme.color("text", scheme))
                .accessibilityAddTraits(.isHeader)
                .padding(.top, 20)
            content()
            if !status.isEmpty {
                Text(status)
                    .centraidType("small")
                    .foregroundStyle(Theme.color("textSoft", scheme))
                    .accessibilityIdentifier("kit-sheet-status")
            }
            if let primary {
                KitInkButton(label: primary.label, destructive: primary.destructive, action: primary.action)
            }
        }
        .padding(.horizontal, CentraidGeometry.pageMargin)
        .padding(.bottom, 16)
        // THE ROOM FILLS ITS SHEET, top-aligned: sized to its content, it
        // floated mid-sheet as a band of `bgElev` between two slabs of the
        // system's own sheet ground.
        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
        .presentationBackground(Theme.color("bgElev", scheme))
        .presentationDragIndicator(.visible)
        .presentationDetents([.medium, .large])
    }
}

/// A sheet's primary verb. **A DESTRUCTIVE ONE IS AN OUTLINE IN `net`**, never
/// a filled red slab (handoff: "the destructive outline"): prominence is earned
/// by consequence, and the consequence is carried by the colour and the
/// sentence, not by shouting.
struct KitInkButton: View {
    let label: String
    var destructive: Bool = false
    let action: () -> Void

    @Environment(\.colorScheme) private var scheme

    var body: some View {
        Button(action: action) {
            Text(label)
                .centraidType("labelOn")
                .foregroundStyle(Theme.color(destructive ? "net" : "onAccent", scheme))
                .frame(maxWidth: .infinity, minHeight: CentraidGeometry.targetMinCoarse)
                .background {
                    if !destructive {
                        RoundedRectangle(cornerRadius: Theme.radius("md", scheme))
                            .fill(Theme.color("accent", scheme))
                    }
                }
                .overlay {
                    if destructive {
                        RoundedRectangle(cornerRadius: Theme.radius("md", scheme))
                            .strokeBorder(Theme.color("net", scheme), lineWidth: 1.5)
                    }
                }
                .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .accessibilityIdentifier(destructive ? "kit-destructive" : "kit-primary")
    }
}
